Add config file

ehelms opened the pull request:
This includes #5

ehelms committed:
Accept config file location via environment variable

This will set the needed configuration options by reading in a
supplied config file and converting what is found there into environment
variables. The expectation is that this configuration file follows
the same format as yggdrasil >0.2.0.

ShimShtein commented on main.go:
```diff
@@ -16,8 +18,27 @@ var yggdDispatchSocketAddr string
-	// Get initialization values from the environment.
 	var ok bool
+
+	configFile, ok := os.LookupEnv("CONFIG_FILE")
+	if ok {
+		config, err := toml.LoadFile(configFile)
+		if err != nil {
+			fmt.Errorf("cannot load config: %w", err)
+			return
+		}
+
+		for _, value := range config.GetArray("env").([]string) {
+			split := strings.Split(value, "=")
+			os.Setenv(split[0], split[1])
+		}
+
+		os.Setenv("YGG_FORWARDER_HANDLER", config.Get("directive").(string))
+	} else {
+		log.Fatal("CONFIG_FILE not set")
+	}
```
Can we reuse the original function from the [yggdrasil package](https://github.com/RedHatInsights/yggdrasil/blob/cdde836c519ac72302714b5f7ae90e45de7c79a7/cmd/yggd/worker.go#L22-L45)? 

At least we should copy the code from there. I think it will be better than developing a new paring algo.

BTW, I see that the original code extracts the directive differently (from the config file name). I think we should stick with that.

ehelms replied:
> Can we reuse the original function from the [yggdrasil package](https://github.com/RedHatInsights/yggdrasil/blob/cdde836c519ac72302714b5f7ae90e45de7c79a7/cmd/yggd/worker.go#L22-L45)?

I don't see this code in 0.2.0 and part of why I opted to implement it ourselves. Keep the coupling to minimum for this one-off thing.

> 
> At least we should copy the code from there. I think it will be better than developing a new paring algo.
> 

Funny thing is, I actually started with that code, but it was not working, getting messy and unwieldy. So I read the docs on the `toml` package we are using and hence the solution I have.

> BTW, I see that the original code extracts the directive differently (from the config file name). I think we should stick with that.

ShimShtein replied:
So let's leave it as is.
The `directive` handling needs to change though.

ehelms replied:
> The `directive` handling needs to change though.

I don't follow this.

ShimShtein replied:
Yggdrasil is [deriving](https://github.com/RedHatInsights/yggdrasil/blob/cdde836c519ac72302714b5f7ae90e45de7c79a7/cmd/yggd/worker.go#L42) the directive from the configuration file name.
We need to do the same, as opposed to reading it directly from the config file (`config.Get("directive").(string)`)

ehelms replied:
Ah ha! Got it thanks

ehelms commented:
@ShimShtein Do you want me to update the environment variable names in this PR? Ideas:

 * WKR_FORWARDER_
 * FORWARDER_
 * WORKER_FORWARDER_

ShimShtein commented:
> Do you want me to update the environment variable names in this PR? 

If you have a couple of spare cycles, it would be awesome. Let's make it `FORWARDER_`.

ehelms commented:
> > Do you want me to update the environment variable names in this PR?
> 
> If you have a couple of spare cycles, it would be awesome. Let's make it `FORWARDER_`.

Handled here -- https://github.com/ShimShtein/yggdrasil-worker-forwarder/pull/7

ShimShtein commented on main.go:
```diff
@@ -16,31 +18,50 @@ var yggdDispatchSocketAddr string
+			os.Setenv(split[0], split[1])
+		}
+
+		os.Setenv("YGG_FORWARDER_HANDLER", strings.TrimSuffix(filepath.Base(configFile), filepath.Ext(configFile)))
```
After the rebase, this should be:
```suggestion
		os.Setenv("FORWARDER_HANDLER", strings.TrimSuffix(filepath.Base(configFile), filepath.Ext(configFile)))
```

ehelms replied:
Done.

ShimShtein reviewed (changes requested):
@ehelms I have merged the parameters renaming, can you please rebase?

ShimShtein committed:
Make CONFIG_FILE optional

We don't have to have the config file, we can specify the parameters directly as env vars. This is what will happen with ygg > 0.2.0

ShimShtein commented:
Merged the PR with a small change: the `CONFIG_FILE` should be optional, since in the newest ygg core, it will be called without that parameter.
